Sources/Brave/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift

```diff
@@ -299,12 +290,19 @@ extension Tab: BraveWalletProviderDelegate {
   
   /// Fetches all allowed accounts for the current origin.
   func getAllowedAccounts(_ type: BraveWallet.CoinType, accounts: [String]) -> [String]? {
+    // For Ethereum, locked status is checked in `EthereumProviderImpl::GetAllowedAccounts`
+    // For Solana, locked status is checked in `SolanaProviderImpl::Connect` before
+    // calling `IsAccountAllowed` delegate method.
     guard let originURL = url?.origin.url,
-          let permissions = Domain.walletPermissions(forUrl: originURL, coin: type) else {
+          let permittedAccountAddresses = Domain.walletPermissions(forUrl: originURL, coin: type) else {
       return []
     }
-    // keyring locked status is checked in `EthereumProviderImpl::GetAllowedAccounts`
-    return permissions
+    // Filter `accounts` requested to return only those permitted
+    return accounts.filter { address in
+      permittedAccountAddresses.contains(where: { permittedAddress in
+        permittedAddress.caseInsensitiveCompare(address) == .orderedSame
+      })
+    }
```

Comment:
not sure why we need filter from `accounts`? isn't this will be the same with `permittedAccountAddresses`?

Reply:
Because core expects us to only return the permitted accounts for the given accounts array they request. For example, consider scenario where a user has 2 accounts (`0xA`, `0xB`) but only permits `0xA` to connect. If they then switch to the non-permitted account `0xB` and try to connect, core will call this `getAllowedAccounts()` function with `accounts: ["0xB"]` but we'd incorrectly return `["0xA"]` is connected so a permission request would not be created.

[Desktop implementation](https://github.com/brave/brave-core/blob/db3af86e921baa371556e158292bc36a74aea794/components/permissions/contexts/brave_wallet_permission_context.cc#L233-L249) of this method.